CST-Group/cst
# Memory storage

### Why was it necessary?

This PR implements Memory Storage, a memory synchronization mechanism. It allows the creation of distributed agents between different CST instances/nodes, and the creation of agents using CST-Java and CST-Python simultaneously.

### How was it done?

The MemoryStorageCodelet has been implemented. By adding it to the agent, the codelet monitors existing memories and synchronizes them with a Redis server when needed.

### Implementation details

In addition to the MemoryStorageCodelet, the MemoryEncoder, which transforms memories to and from Maps with JSON info, and the LamportTime, which implements a logical Lamport clock, were also implemented.

The full communication protocol details is provided in: https://h-iaac.github.io/CST-Python/_build/html/src/Memory%20Storage.html

### How to test?

The Tests test communication between local nodes in Java. To run them, you need to have a Redis server available, which has been added to the test CI. If it is not available, the test is skipped.

A test was also implemented to verify the operation of other Memory Storage instances, which at the moment must be executed manually. The test was successfully executed with MS-Python.

### Future works

Some more robust memory information encoding and typing system is still needed, allowing objects of arbitrary types to be synchronized.
